### PR TITLE
Make brain tests pass

### PR DESCRIPTION
**Pull request by andreas-kupries**
Ref
  - https://jira.suse.com/browse/CAP-526
  - https://github.com/SUSE/kubecf/issues/29

This draft will contain accumulated changes to the release to make brain tests pass.

See also the controlling PR https://github.com/SUSE/kubecf/pull/242

### CODE REVIEW

File: packages/helm/packaging [hunk 1]
```diff
@@ -15,8 +14,12 @@ do
...
     DEST="${BIN_DIR}/${helm_version}"
     mkdir -p "${DEST}"
-    helm_url="https://storage.googleapis.com/kubernetes-helm/helm-v${helm_version}-linux-amd64.tar.gz"
-    curl -L "${helm_url}" | tar zx --strip-components 1 -C "${DEST}"
+    helm_url="https://get.helm.sh/helm-v${helm_version}-linux-amd64.tar.gz"
+    echo URL: ${helm_url}
+    curl -L "${helm_url}" > helm-tarball
+    ls -l helm-tarball
+    file  helm-tarball
+    cat   helm-tarball | tar zx --strip-components 1 -C "${DEST}"
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply packages/helm/packaging]: The tracing helping is actually about the net/http in the docker uploader.
I will see about removing this.

File: src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect [hunk 2]
```diff
@@ -3,18 +3,48 @@
...
+# path `resources/cache` requires depth 3. If this leads to
+# misdetections for `dependencies` in the children of the BP toplevels
+# we will have to split the `find` command into two. For now we do not
+# care.
+
+LOCATION="/tmp/buildpacks"
+
+report() {
+    buildpack="${1}"
     echo "### Buildpack cache detection: ERROR: found cached dependencies in ${buildpack}"
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect]: Trying to distribute offline buildpacks is legal trouble, so error. AFAIK

File: src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect [hunk 3]
```diff
@@ -3,18 +3,48 @@
...
+	    buildpack_dir="$(dirname ${buildpack})"
+	    language="$(grep language ${buildpack_dir}/manifest.yml)"
+	    ;;
+    esac
     echo "### Buildpack cache detection: ${language}"
-  done
-  exit 1
+}
+
+offenders="$(find "${LOCATION}" -maxdepth 3 -type d -path '*/resources/cache' -o -name dependencies)"
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect]: It is more explicit about the paths. Will trial. I should have the offline buildpacks still around to test against.

File: packages/helm/packaging (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply packages/helm/packaging]: Addressed with commit 79bc424813e086a01dae04db0f0bfb5fe2a9f04d. Tested to work.

File: src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect]: Addressed with commit 79bc424813e086a01dae04db0f0bfb5fe2a9f04d. Tested to work.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb [hunk 4]
```diff
@@ -70,38 +73,72 @@
...
+sleep 5
+
+REGISTRIES.each_pair do |regname, registry_url|
+    puts "wait for registry to be available ......................................... #{regname}"
     # Wait for the registry to be available
     run_with_retry 60, 1 do
         run "curl -kv #{registry_url}/v2/"
     end
+end
+
```
Comment (mook-as) [inline src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Hmm, found it useful for me to to curl `http://uploader.#{ENV['CF_DOMAIN']}` until that responds with `HTTP 200` too; not sure, though, might just be something local?

File: src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb [hunk 5]
```diff
@@ -10,6 +10,11 @@
 
 use_global_timeout
 
+puts "#{c_bold}Test disabled.#{c_reset}"
+puts "#{c_bold}Depends on CAP-1293#{c_reset} to enable NFS persi support."
```
Comment (mook-as) [inline src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb]: Do we have a GitHub reference here? (Jira is not a good place to keep things with a public impact.)

File: src/acceptance-tests-brain/test-scripts/013_credhub_test.rb [hunk 6]
```diff
@@ -8,8 +8,18 @@
 CH_CLI = 'credhub'
 CH_SERVICE = "https://credhub.#{ENV['CF_DOMAIN']}"
 
-# Check if credhub is running, otherwise skip the test.
-status = run_with_status('curl', '--silent', '--fail', '--insecure', "#{CH_SERVICE}/info")
+# Check if credhub is running, otherwise skip the test.  We do not
+# break on the first failure, but check a bit more, in case it was a
+# transient issue with the endpoint. We do accept sucess immediately,
+# tough.
```
Comment (mook-as) [inline src/acceptance-tests-brain/test-scripts/013_credhub_test.rb]:
```suggestion
# though.
```
… though it _is_ to amusing to think that you mean "we accept it, whether you like it or not" ;)

File: src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect [hunk 7]
```diff
@@ -3,18 +3,43 @@
...
+LOCATION="/home/andreask/fast/dev/kubecf-1/_work/offline-bp"
+
+report() {
+    buildpack="${1}"
     echo "### Buildpack cache detection: ERROR: found cached dependencies in ${buildpack}"
-    buildpack_dir=`dirname ${buildpack}`
-    language=`grep language ${buildpack_dir}/manifest.yml`
+    if test -d "${buildpack}/resources/cache"
+    then
+	language="language: java"
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect]: sure

File: src/acceptance-tests-brain/test-resources/docker-uploader/manifest.yml [hunk 8]
```diff
@@ -5,8 +5,8 @@ shared: &shared
   command: ./bin/registry serve ./config.yml
   env: &env
     REGISTRY_HTTP_ADDR: :8080
-    REGISTRY_LOG_ACCESSLOG_DISABLED: true
-    REGISTRY_LOG_LEVEL: error
+    REGISTRY_LOG_ACCESSLOG_DISABLED: false
+    REGISTRY_LOG_LEVEL: debug
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/manifest.yml]: possibly not. OTOH, I like to have logs of problems available immediately when something breaks, and not after another iteration where the logs are activated. Given the iteration time I really would like to keep this.

File: src/acceptance-tests-brain/test-resources/resources.csv [hunk 9]
```diff
@@ -1,2 +1 @@
-cf-plugin-usb,https://github.com/SUSE/cf-usb-plugin/releases/download/1.0.0/cf-usb-plugin-1.0.0.0.g47b49cd-linux-amd64
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/resources.csv]: `cf-usb` was stricken completely, per Troy. Even before it ran into trouble with kube 1.16.

File: src/acceptance-tests-brain/test-scripts/009_backup_test.rb [hunk 10]
```diff
@@ -23,6 +23,8 @@
     end
 end
 
+puts "Set up users and org/space roles ..."
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/009_backup_test.rb]: Technically not. It made it easier for me to see where each command (output) came from.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb [hunk 11]
```diff
@@ -25,9 +25,9 @@
 run "cf bind-security-group #{CF_SEC_GROUP} #{$CF_ORG} #{$CF_SPACE} --lifecycle staging"
 run "cf bind-security-group #{CF_SEC_GROUP} #{$CF_ORG} #{$CF_SPACE} --lifecycle running"
 
+# Defered: 'secure-registry'   => "https://secure-registry.#{ENV['CF_DOMAIN']}",          # Router SSL cert
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: No, not yet.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb [hunk 12]
```diff
@@ -70,38 +73,72 @@
...
+
+Thread.new{
+    run "cf logs uploader > #{tmpdir}/app-uploader.log"
+}.abort_on_exception = true
+
 REGISTRIES.each_pair do |regname, registry_url|
+    Thread.new{
+        run "cf logs #{regname} > #{tmpdir}/app-#{regname}.log"
+    }.abort_on_exception = true
+end
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Sure

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb (hunk 4 above)
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: In the interest of robustness I would say, let us add that.

File: src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb (hunk 5 above)
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb]: Not aware of any (I looked into the CAP, just rechecked, no such link)

File: src/acceptance-tests-brain/test-scripts/013_credhub_test.rb (hunk 6 above)
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/013_credhub_test.rb]: Heh. Will fix. The `sucess` above as well, should `success`.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Found ticket made by @gak: https://github.com/cloudfoundry-incubator/kubecf/issues/382
Added to notice.

[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: :man_facepalming: Wrong section, this is about persi, not secure registry.

File: src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb (hunk 5 above)
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb]: Found ticket made by @gak: cloudfoundry-incubator/kubecf#382
Added to notice, and CAP ticket.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Created https://github.com/cloudfoundry-incubator/kubecf/issues/466

File: src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect]: Addressed with commit f2be8c95181e3a6e5e099501d59df6ae6a4d7c48.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb (hunk 4 above)
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Addressed with commit f2be8c95181e3a6e5e099501d59df6ae6a4d7c48.

File: src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb (hunk 5 above)
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/011_nfspersi_test.rb]: Addressed with commit f2be8c95181e3a6e5e099501d59df6ae6a4d7c48.

File: src/acceptance-tests-brain/test-resources/docker-uploader/main.go [hunk 13]
```diff
@@ -192,65 +201,158 @@ func makeLayer() (resultFile *os.File, err error) {
 	return outFile, nil
 }
 
+func trace (req *http.Request) *http.Request {
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/main.go]: Addressed with commit f2be8c95181e3a6e5e099501d59df6ae6a4d7c48.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Addressed with commit f2be8c95181e3a6e5e099501d59df6ae6a4d7c48.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Addressed with commit f2be8c95181e3a6e5e099501d59df6ae6a4d7c48.

**Comment by andreas-kupries** [comment]
My last run of the tests, with all the comments addressed yielded

```
  Failed tests:
    016_minibroker_mongodb_test.rb with exit code 1
    004_tcprouting_test.rb with exit code 1
    014_minibroker_mariadb_test.rb with exit code 1
    015_minibroker_postgres_test.rb with exit code 1
```

None of the tests touched by the changes failed. I suspect that it was something transient about these, independent of the changes.

So, not seeing it as blocker to merging, also not adverse to an independent last run by you (@mook-as ) before doing so.

File: src/acceptance-tests-brain/test-scripts/013_credhub_test.rb (hunk 6 above)
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/013_credhub_test.rb]: Addressed by commit f2be8c95181e3a6e5e099501d59df6ae6a4d7c48

File: src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect [hunk 14]
```diff
@@ -3,18 +3,47 @@
...
+#
+# This means that a directory like
+# `vendor/bundle/ruby/2.6.0/gems/activesupport-4.2.11.1/lib/active_support/dependencies`
+# as found in the online PHP buildpack does NOT count.
+#
+# This is achieved by explicitly iterating over the buildpacks and
+# testing for the problematic directories.
+
+LOCATION="/tmp/buildpacks"
+LOCATION="/home/andreask/fast/dev/kubecf-1/_work/offline-bp"
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect]: Grrk. Will remove.

File: src/acceptance-tests-brain/test-resources/docker-uploader/main.go [hunk 15]
```diff
@@ -218,37 +220,52 @@ func uploadBlob(registry, name string) (string, error) {
...
 		newURL, err := url.Parse(resp.Header.Get("Location"))
 		if err != nil {
 			return "", err
 		}
 		query := newURL.Query()
 		query.Add("digest", "sha256:"+digest)
 		newURL.RawQuery = query.Encode()
-		req, err = http.NewRequest(http.MethodPut, newURL.String(), file)
+		newreq, err := http.NewRequest(http.MethodPut, newURL.String(), file)
+		// file = io.Reader - body.
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/main.go]: Will fix.

File: src/acceptance-tests-brain/test-resources/docker-uploader/main.go [hunk 16]
```diff
@@ -218,37 +220,52 @@ func uploadBlob(registry, name string) (string, error) {
...
 	req.Header.Set("Content-Length", fmt.Sprintf("%d", fileLength))
 	req.Header.Set("Content-Type", "application/octet-stream")
 	resp, err := http.DefaultClient.Do(req)
+
 	if err != nil {
 		return "", err
 	}
 
 	for resp.StatusCode == http.StatusAccepted {
+		//
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/main.go]: Most likely left over from some removal. Will remove.

File: src/acceptance-tests-brain/test-resources/docker-uploader/manifest.yml [hunk 17]
```diff
@@ -34,5 +34,7 @@ applications:
 
 - name: uploader
   buildpack: binary_buildpack
+  env:
+    GODEBUG: http2client=0
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/manifest.yml]: +1

File: src/acceptance-tests-brain/test-resources/docker-uploader/config.yml [hunk 18]
```diff
@@ -16,3 +16,6 @@ health:
     enabled: true
     interval: 10s
     threshold: 3
+compatibility:
+  schema1:
+    enabled: true
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/config.yml]: +1

[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/config.yml]: Addressed with commit bd60882417a962faa6f1a97f3f2e6512b897a152.

File: src/acceptance-tests-brain/test-resources/docker-uploader/manifest.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/manifest.yml]: Addressed with commit bd60882417a962faa6f1a97f3f2e6512b897a152.

File: src/acceptance-tests-brain/test-resources/docker-uploader/main.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/main.go]: Addressed with commit bd60882417a962faa6f1a97f3f2e6512b897a152.

File: src/acceptance-tests-brain/test-resources/docker-uploader/main.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/docker-uploader/main.go]: Addressed with commit bd60882417a962faa6f1a97f3f2e6512b897a152.

File: src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect (hunk 14 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-resources/buildpack_inspector_buildpack/bin/detect]: Addressed with commit bd60882417a962faa6f1a97f3f2e6512b897a152.

File: src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb [hunk 19]
```diff
@@ -4,104 +4,155 @@
...
+  REGISTRIES.each_pair do |regname, registry_url|
+    puts "wait for registry to be available ......................................... #{regname}"
     # Wait for the registry to be available
     run_with_retry 60, 1 do
-        run "curl -kv #{registry_url}/v2/"
+      run "curl -kv #{registry_url}/v2/"
     end
+  end
+  run_with_retry 60, 1 do
+    run "curl -kv http://uploader.#{ENV['CF_DOMAIN']}"
```
Comment (mook-as) [inline src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: This doesn't have `--fail`; does that actually wait until it's up?

Comment (andreas-kupries) [reply src/acceptance-tests-brain/test-scripts/010_insecure_registry_test.rb]: Added `--fail` in commit b86a1aea256d52838f3584c756a75dfe0b8147c5.
From the quick look at the manpage this does seem to only suppress output and prevent conversion of errors, not be needed to have errors at all. To wit

```
Fail silently (no output at all) on server errors. This is mostly done to better enable scripts etc to better deal with failed attempts. In normal cases when an HTTP server fails to deliver a document, it returns an HTML document stating so (which often also describes why and more). This flag will prevent curl from outputting that and return error 22.
```

**Comment by andreas-kupries** [comment]
> Turns out my problem is that tcp-router-public has a health check port (80),
> which is clobbering the router-public port (80).

How can this be a problem on your machine, yet not on mine ?
Are you applying custom configs when you start either minikube or kubecf ?
(I know that I don't).

**Comment by mook-as** [comment]
> How can this be a problem on your machine, yet not on mine ?

I believe _which_ service (`router` or `tcp-router`) gets the port depends on timing.  I've just checked again that this is the case…

> Are you applying custom configs when you start either minikube or kubecf ?

Not very interesting ones:
<details>
<summary> <tt>values.yaml</tt></summary>

```yaml
system_domain: 192.168.39.209.omg.howdoi.website
services:
  router:
    type: ClusterIP
    externalIPs: ["192.168.39.209"]
  ssh-proxy:
    type: ClusterIP
    externalIPs: ["192.168.39.209"]
  tcp-router:
    type: ClusterIP
    externalIPs: ["192.168.39.209"]
  port_range:
      start: 20001
      end:   20010
features:
  credhub:
    enabled: false
```
</details>
(Plus enabling the tests.)

And for minikube:

<details>
<summary> <tt>minikube config view</tt> </summary>

```
- coredns: false
- heapster: true
- kube-dns: true
- memory: 16384
- vm-driver: kvm2
- WantReportErrorPrompt: false
- WantUpdateNotification: false
- addon-manager: true
```
</details>

I think my problem can be moved to a new issue, though.

**Comment by andreas-kupries** [comment]
My `minikube config view` reports
```
- WantReportErrorPrompt: false
- heapster: true
- vm-driver: kvm
```

My `values.yaml` is mostly the same.
- I am not disabling credhub, nor am I setting the `port_range`.
- I am using `LoadBalancer` over `ClusterIP` however.
Maybe because I am not using ingress.
So it may be ingress which makes the difference ?

**Comment by mook-as** [comment]
I'm not using ingress either. I used `ClusterIP` over `LoadBalancer` because I thought the latter didn't work with minikube.  I'll check with LB, then.
In the mean time, I suppose we should merge?

**Comment by andreas-kupries** [comment]
@mook-as I am ok with merging.

**Pull request merged**